clients/src/main/java/org/apache/kafka/clients/admin/AdministrativeClient.java

```diff
@@ -0,0 +1,628 @@
+     */
+    private <T> T waitForResultFromController(String requestName,
+                FutureProvider<RequestFuture<T>> futureProvider) throws InterruptedException {
+        long startMs = time.milliseconds();
+        long remaining = requestTimeoutMs;
+        Node controller = metadata.fetch().controller();
+        do {
+            if (controller == null) {
+                log.debug("Requesting metadata update in order to find controller id.");
+                consumerClient.awaitMetadataUpdate(remaining);
+                controller = metadata.fetch().controller();
+                log.debug("New controller id is " + controller);
+            } else {
+                RequestFuture<T> future = futureProvider.provide(controller);
+                consumerClient.poll(future, remaining);
+                if (future.isDone()) {
+                    if (future.succeeded())
+                        return future.value();
+                    else if (!future.isRetriable())
+                        throw future.exception();
+                }
+            }
+            remaining = time.milliseconds() + requestTimeoutMs - startMs;
+            log.debug("waitForResult(" + requestName + "): " + remaining + " ms remaining.");
+        } while (remaining > 0);
+        throw new TimeoutException(requestName + " timed out after " + requestTimeoutMs + " ms");
+    }
+
+    public enum CreateTopicsFlags {
+        /**
+         * Return immediately without waiting for the topic to be created.
+         */
+        NONBLOCKING,
+
+        /**
+         * Perform validation that the proposed topics could be created, but do not actually create them.
+         */
+        VALIDATE_ONLY;
+    }
+
+    /**
+     * Create a batch of topics.
+     *
+     * @param newTopics                 A list of new topics to create.
+     * @param flags                     The flags to use on the requests.
+     * @return                          A map from topic names to exceptions.  If the entry in the map
+     *                                  is null, the topic has been successfully created.
+     * @throws InterruptedException     If the thread receives an InterruptedException before finishing.
+     * @throws TimeoutException         If the thread times out before learning which node is the controller.
+     *                                  Individual topic creations may also time out, causing exceptions to
+     *                                  be stored in the map.
+     */
+    public Map<String, KafkaException> createTopics(final List<NewTopic> newTopics,
+            EnumSet<CreateTopicsFlags> flags) throws InterruptedException {
+        final boolean validateOnly = flags.contains(CreateTopicsFlags.VALIDATE_ONLY);
+        final int timeout = flags.contains(CreateTopicsFlags.NONBLOCKING) ? 0 : requestTimeoutMs;
+        final Map<String, TopicDetails> topicMap = new HashMap<>();
+        for (NewTopic newTopic: newTopics) {
```

Comment:
I didn't make it a set because I didn't want to have to worry about hashCode and equals.  I think you're right; it should just be a Collection.